Enable full Encodings on Windows Console by Default

tarekgh opened the pull request:
by default .NetCore doesn't enable the codepage encoding and the app has to opt-in to the full encoding through the call
	Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
This was not a good stand for the Console specially on Windows as the Console always using encoding which is not supported by default and we had to fallback using UTF8.
In this change we are supporting the full OS encoding on Console and System.Diagnostics.Process without bringing the big encoding size cost (which is 600+ KB). The way we support it is we create an OSEncoding class wrapper which will call the OS instead of using our own encoding data.
The only restriction we’ll have is cannot use encoding fallback with OSEncoding as this will be very slow and not intuitive. But this is acceptable as usually apps don’t use the console encodings for such fallback operation. But if anyone want to do that can still register the encoding provider and get full support.

tarekgh commented:
@jkotas PTAL

cc @danmosemsft @weshaggard @nguerrera

tarekgh commented:
#14496

jkotas commented on src/System.Console/src/System.Console.csproj:
```diff
@@ -49,6 +46,15 @@
+      <Link>Common\System\Text\OSEncoding.Windows.cs</Link>
+    </Compile>
+    <Compile Include="$(CommonPath)\System\Text\EncodingTable.cs">
+      <Link>Common\System\Text\EncodingTable.cs</Link>
```
We should not be bundling multiple copies of these tables just to return the proper encoding name that are very unlikely to be ever called.

Instead, the name properties should be implemented by getting the global encoding for the given codepage via `Encoding.GetEncoding` and forwarding the request to it. If you do not register the provider, you will get an exception telling you to register the full encoding provider to get the encoding name - you will be only able to get the names of these encodings if you register the full encoding provider.

jkotas commented on src/Common/src/System/Text/EncodingHelper.Windows.cs:
```diff
@@ -38,6 +38,12 @@ internal static Encoding GetSupportedConsoleEncoding(int codepage)
                 {
                 }
```
We should not need the try/catch above anymore.

jkotas commented on src/Common/src/System/Text/OSEncoding.Windows.cs:
```diff
@@ -0,0 +1,228 @@
+
+            fixed (byte* pBytes = bytes)
+            {
+                return MultiByteToWideChar(pBytes+index, count, null, 0);
```
I do not think that this will work with straddling bytes.

tarekgh commented:
@jkotas I have added the encoder (to handle the surrogate cases) and decoder for DBCS codepages. this should cover the majority of the supported encodings. we still have a couple of cases that I'll investigate and handle it later after the holidays. thanks for reviewing it.

jkotas commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,262 @@
+            if (byteCount == 0)
+                return 0;
+
+            char lastChar = charCount > 0 && !flush && Char.IsHighSurrogate(chars[charCount - 1]) ? chars[charCount - 1] : NULL_CHAR;
```
Does this work correctly if the source `chars` array contains (char)0?

tarekgh replied:
yes it should work correctly even the source array contains nulls. lastChar is kind of flag either has upper surrogate or null. if it is null, we don't use it at all in the encoding conversion.

---
In reply to: [93727921](https://github.com/dotnet/corefx/pull/14649#discussion_r93727921) [](ancestors = 93727921)

jkotas replied:
Make sense.

jkotas commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,262 @@
+            if (charCount == 0 && lastChar == NULL_CHAR && !flush)
+                return 0;
+
+            char [] bufferToProcess = new char[charCount+1];
```
This copy looks pretty expensive. It would be better to create a small buffer on stack with the two characters, call WideCharToMultiByte on it, and then call WideCharToMultiByte on the original buffer.

jkotas commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,262 @@
+            fixed (char* pChars = chars)
+            fixed (byte* pBytes = bytes)
+            {
+                char *pSingleElementArray = stackalloc char[1]; // to avoid passing null to Convert
```
No need to use stackalloc - `char dummyChar` is better.

jkotas commented on src/Common/src/Interop/Windows/kernel32/Interop.GetCPInfoEx.cs:
```diff
@@ -8,7 +8,7 @@ internal partial class Interop
 {
     internal partial class Kernel32
     {
-        private const uint CP_ACP = 0x0u;
```
`CP_ACP` is the Windows SDK name for this. It should stay that way.

The convention is to use the same name as what's used in the native code (https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md).

jkotas commented on src/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs:
```diff
@@ -43,22 +43,12 @@ public void TestChangesInConsoleEncoding()
 
             try
```
We should add proper test coverage for this.

tarekgh replied:
The current test already testing our change without need to register the encoding provider. what exactly the case you are asking to test?

---
In reply to: [93728474](https://github.com/dotnet/corefx/pull/14649#discussion_r93728474) [](ancestors = 93728474)

jkotas replied:
The current console tests have coverage for a simple ASCII strings only: https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Console/tests/SyncTextReader.cs#L15

This change has a number of interesting paths that handle non-ASCII strings.

tarekgh replied:
ok, will try to add cases for non-ASCII considering setting the codepage on the console.

tarekgh commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,262 @@
+            fixed (char* pChars = chars)
+            fixed (byte* pBytes = bytes)
+            {
+                char *pSingleElementArray = stackalloc char[1]; // to avoid passing null to GetByteCount
```
I'll change that

---
In reply to: [93727715](https://github.com/dotnet/corefx/pull/14649#discussion_r93727715) [](ancestors = 93727715)

tarekgh commented:
thanks again @jkotas 
I have done all fixes according to your comments.

tarekgh commented:
@jkotas did you have a chance to look at the changes?

jkotas commented:
The implementation of GetSupportedConsoleEncoding does not look right. Why it cannot be just - what does the extra check for defaultEncCodePage achieves?

```
        internal static Encoding GetSupportedConsoleEncoding(int codepage)
        {
            if (codepage != Encoding.UTF8.CodePage)
            {
                return new OSEncoding(codepage);
            }

            return new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
        }
```

jkotas commented:
The behavior around surrogates and invalid surrogates accross straddling boundaries does not look consistent. E.g.

```
static void Main()
{
    string x = "\uD800";
    string y = "b";
    Console.Write(x); Console.Write(y); Console.WriteLine();
    Console.WriteLine(x + y);
}
```

Actual result:

b
?b

Expected result:

?b
?b

tarekgh commented:
> The implementation of GetSupportedConsoleEncoding does not look right. Why it cannot be just - what does the extra check for defaultEncCodePage achieves?

by default we support other encodings like UTF16, ASCII...etc.s which we should try to create it first from the supported encoding and not from OSEncoding. Also if the encoding provider is registered then we should use this instead of OSEncoding for the sake of the performance too. I believe the current implementation I have is correct.

jkotas commented:
I see now ... why is it ok to remove the try/catch block that used to be there?

tarekgh commented:
> I see now ... why is it ok to remove the try/catch block that used to be there?

usually we return UTF8 as the default encoding if the provider is not registered. if we got some other value different than UTF8 that means we have a provider registered. we are keeping the other condition defaultEncCodePage == codepage because if someone changed the default on the system to one of the defautl supported codepage like UTF16, ASCII...etc. which should work too and shouldn't throw exception.

tarekgh commented:
@jkotas 

> The behavior around surrogates and invalid surrogates accross straddling boundaries does not look consistent. E.g.

this was a corner case when the buffer had one high surrogate character but it is a good catch. I have fixed the case and also added a test case to handle this case.

stephentoub commented on src/Common/src/Interop/Windows/kernel32/Interop.GetCPInfoEx.cs:
```diff
@@ -24,18 +22,37 @@ private unsafe struct CPINFOEXW
+                for (int i=0; i<10; i+=2)
+                {
+                    if (leadyByteRanges[i] == 0)
+                        break;
```
Nit: could be:
```
for (int i = 0; i < 10 && leadByteRanges[i] != 0; i += 2)
```

stephentoub commented on src/Common/src/Interop/Windows/kernel32/Interop.GetCPInfoEx.cs:
```diff
@@ -24,18 +22,37 @@ private unsafe struct CPINFOEXW
             internal fixed byte CodePageName[260];
         }
 
+        internal static unsafe int GetLeadByteRanges(int codePage, byte[] leadyByteRanges)
```
Nit: leadyByteRanges => leadByteRanges

stephentoub commented on src/Common/src/Interop/Windows/kernel32/Interop.GetCPInfoEx.cs:
```diff
@@ -24,18 +22,37 @@ private unsafe struct CPINFOEXW
+            return count;
+        }
+
         internal static bool TryGetACPCodePage(out int codePage)
```
Nit: stylistically, I'd have shortened this a bit, e.g.
```C#
internal static unsafe bool TryGetACPCodePage(out int codePage)
{
    CPINFOEXW cpInfo;
    if (GetCPInfoExW(CP_ACP, 0, &cpInfo) != 0)
    {
        codePage = (int)cpInfo.CodePage;
        return true;
    }
    else
    {
        codePage = 0;
        return false;
    }
}
```

stephentoub commented on src/Common/src/System/Text/DBCSDecoder.cs:
```diff
@@ -0,0 +1,293 @@
+
+namespace System.Text
+{
+    internal class DecoderDBCS : Decoder
```
Decoder is serializable, but this new DecoderDBCS is not marked [Serializable].  Is there any situation where the Decoder instance previously used in whatever situation this is being used now could have been handed out to user code and serialized?

stephentoub replied:
Nit: sealed

stephentoub commented on src/Common/src/System/Text/DBCSDecoder.cs:
```diff
@@ -0,0 +1,293 @@
+{
+    internal class DecoderDBCS : Decoder
+    {
+        private Encoding _encoding;
```
Nit: readonly

stephentoub commented on src/Common/src/System/Text/DBCSDecoder.cs:
```diff
@@ -0,0 +1,293 @@
+    internal class DecoderDBCS : Decoder
+    {
+        private Encoding _encoding;
+        private byte[] _leadyByteRanges = new byte[10]; // Max 5 ranges
```
Nit: readonly

stephentoub replied:
Nit: leady => lead

stephentoub commented on src/Common/src/System/Text/DBCSDecoder.cs:
```diff
@@ -0,0 +1,293 @@
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            bool excludeLastByte = count > 0 || (!flush && IsLastByteALeadByte(bytes, count));
```
If count == 0 && flush is false, we end up calling IsLasByteALeadByte with bytes and 0, and then IsLastByteALeadByte ends up accessing bytes[-1].  Is that ok?

stephentoub commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,252 @@
+
+namespace System.Text
+{
+    internal class OSEncoder : Encoder
```
Same serialization question

stephentoub replied:
Nit: sealed

stephentoub commented on src/Common/src/System/Text/OSEncoder.cs:
```diff
@@ -0,0 +1,252 @@
+    internal class OSEncoder : Encoder
+    {
+        private char _charLeftOver;
+        private Encoding _encoding;
```
Nit: readonly

stephentoub commented on src/Common/src/System/Text/OSEncoding.Windows.cs:
```diff
@@ -0,0 +1,258 @@
+
+namespace System.Text
+{
+    internal class OSEncoding : Encoding
```
Nit: sealed

stephentoub commented on src/Common/src/System/Text/OSEncoding.Windows.cs:
```diff
@@ -0,0 +1,258 @@
+{
+    internal class OSEncoding : Encoding
+    {
+        private int _codePage;
```
Nit: readonly

stephentoub commented on src/System.Console/tests/ConsoleEncoding.cs:
```diff
@@ -0,0 +1,66 @@
+    public void TestEncoding(string inputString)
+    {
+        var outConsoleStream = Console.Out;
+        var inConsoleStream = Console.In;
```
Nit: var => StreamWriter/Reader

stephentoub commented on src/System.Console/tests/ConsoleEncoding.cs:
```diff
@@ -0,0 +1,66 @@
+
+            string outString;
+            using (MemoryStream ms = new MemoryStream(inputBytes, false))
+            {
```
Nit: unnecessary braces/indentation

tarekgh commented:
@stephentoub thanks for the comments. I am going to submit a new PR addressing your comments.
